src/ring_theory/congruence.lean

```diff
@@ -271,4 +276,167 @@ def mk' [non_assoc_semiring R] (c : ring_con R) : R →+* c.quotient :=
 
 end quotient
 
+/-! ### Lattice structure
+
+The API in this section is copied from `group_theory/congruence.lean`
+-/
+
+section lattice
+variables [has_add R] [has_mul R]
+
+/-- For congruence relations `c, d` on a type `M` with multiplication and addition, `c ≤ d` iff
+`∀ x y ∈ M`, `x` is related to `y` by `d` if `x` is related to `y` by `c`. -/
+instance : has_le (ring_con R) := ⟨λ c d, ∀ ⦃x y⦄, c x y → d x y⟩
+
+/-- Definition of `≤` for congruence relations. -/
+theorem le_def {c d : ring_con R} : c ≤ d ↔ ∀ {x y}, c x y → d x y := iff.rfl
+
+/-- The infimum of a set of congruence relations on a given type with multiplication and
+addition. -/
+instance : has_Inf (ring_con R) :=
+⟨λ S,
+  { r := λ x y, ∀ c : ring_con R, c ∈ S → c x y,
+    iseqv := ⟨λ x c hc, c.refl x, λ _ _ h c hc, c.symm $ h c hc,
+              λ _ _ _ h1 h2 c hc, c.trans (h1 c hc) $ h2 c hc⟩,
+    add' := λ _ _ _ _ h1 h2 c hc, c.add (h1 c hc) $ h2 c hc,
+    mul' := λ _ _ _ _ h1 h2 c hc, c.mul (h1 c hc) $ h2 c hc }⟩
+
+/-- The infimum of a set of congruence relations is the same as the infimum of the set's image
+    under the map to the underlying equivalence relation. -/
+lemma Inf_to_setoid (S : set (ring_con R)) : (Inf S).to_setoid = Inf (to_setoid '' S) :=
+setoid.ext' $ λ x y, ⟨λ h r ⟨c, hS, hr⟩, by rw ←hr; exact h c hS,
+  λ h c hS, h c.to_setoid ⟨c, hS, rfl⟩⟩
+
+/-- The infimum of a set of congruence relations is the same as the infimum of the set's image
+    under the map to the underlying binary relation. -/
+lemma Inf_def (S : set (ring_con R)) :
+  ⇑(Inf S) = Inf (@set.image (ring_con R) (R → R → Prop) coe_fn S) :=
+by { ext, simp only [Inf_image, infi_apply, infi_Prop_eq], refl }
+
+instance : partial_order (ring_con R) :=
+{ le := (≤),
+  lt := λ c d, c ≤ d ∧ ¬d ≤ c,
+  le_refl := λ c _ _, id,
+  le_trans := λ c1 c2 c3 h1 h2 x y h, h2 $ h1 h,
+  lt_iff_le_not_le := λ _ _, iff.rfl,
+  le_antisymm := λ c d hc hd, ext $ λ x y, ⟨λ h, hc h, λ h, hd h⟩ }
+
+/-- The complete lattice of congruence relations on a given type with multiplication and
+addition. -/
+instance : complete_lattice (ring_con R) :=
```

Comment:
Is that true? My bet is "probably".
```suggestion
instance : complete_distrib_lattice (ring_con R) :=
```

Reply:
Is it true on `setoid`? Either way, this instance is just copied from the instance for `con M` and `add_con M`; there no novel generalization going on in this PR, it's just mechanistic copying.

Reply:
Do you mind changing `group_theory.congruence` then? This PR needs forward-porting anyway.